elastic/beats
# Refactor kubernetes autodiscover to avoid skipping short-living pods

## What does this PR do?

Refactor logic in kubernetes autodiscover that decides when to generate events to try to address #22718.

Kubernetes autodiscover can generate events without network information now (without `host` or `port`/`ports`). This allows to generate events for pods that haven't started yet, or have succeeded/failed before generating a running event. These events still include the container id, so they can be used to collect logs. Still, no start event is generated if no pod ip and no container ids are available.

Some helpers have been added to obtain relevant information from pods and their containers.

Some additional small refactors are done to improve readability.

## Why is it important?

Current logic is checking similar things at the pod and container levels, try to simplify this logic focusing in containers only.

No matter what is the state of the pod, if there is a container running or trying to run, even if it is unhealthy, some configuration should be generated, so logs can be collected.

## Checklist



- [x] I have added tests that prove my fix is effective or that my feature works. But more tests would be needed.

## Author's Checklist


- [x] Fix update so it definitely stops configurations, respecting the cleanup timeout.
- [x] Double-check that nothing breaks with events without `data.host`.
  This is logged at the debug level:
  ```
  2021-04-08T12:53:34.453+0200	DEBUG	[autodiscover]	template/config.go:156	Configuration template cannot be resolved: field 'data.host' not available in event or environment accessing 'hosts' (source:'/home/jaime/tmp/metricbeat-autodiscover-k8s.yml')
  ```
- [x] Actually check that the refactor solves the issues and doesn't break anything else.
- [x] Try to add more testing. A couple of cases added to current tests.
- [x] Check that nothing breaks with hints including `data.host` or `data.ports`.
  - [x] `data.ports.*` doesn't seem to be working, check if this is a regression.  Named ports were not added to container events. They are now.
  - [x] Multiple configurations generated for pods with multiple containers, expected? compare with master. False alarm.
- [x] Check autodiscover with heartbeat.

## Author's notes for the future

- When using autodiscover templates, if the condition matches pod conditions, container events will match too because they also include pod metadata. In pods with multiple containers this may lead to events with metadata of the incorrect container. This could be possibly avoided by matching per pod metadata only in the pod events. This is also happening in master, not solving in this PR. 
- `data.ports` is not included in container-level events, so it doesn't work when conditions for specific containers are used.

## How to test this PR locally

* Check that everything keeps running with some autodiscover happy cases.
* Check that logs are collected from a pod in CrashLoopBackOff (`kubectl run --image=redis redis -- echo foo`).

## Related issues


- Closes #22718.

## Use cases


Collect logs from containers in short-living or failing pods.
